hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,214 @@
+                currentOperationData.invocation = null;
+                continue;
+            }
+
+            long duration = nowNanos - currentOperationData.startNanos;
+            if (duration < slowOperationThresholdNanos) {
+                continue;
+            }
+
+            SlowOperationLog.Invocation invocation = currentOperationData.invocation;
+            if (invocation != null) {
+                // update existing invocation to avoid creation of stack trace
+                invocation.duration = TimeUnit.NANOSECONDS.toMillis(duration);
+                invocation.lastAccess = nowNanos;
+                continue;
+            }
+
+            // create the stack trace once for this operation and cache the invocation for upcoming updates
+            SlowOperationLog log = getOrCreateSlowOperationLog(operationRunners[i], currentOperation);
+            currentOperationData.invocation = log.getOrCreateInvocation(operationHashCode, duration, nowNanos, nowMillis);
+
+            // log a warning and print the full stack trace each 100 invocations
+            int totalInvocations = log.getTotalInvocations();
+            if (totalInvocations == 1) {
+                LOGGER.warning(format("Slow operation detected: %s%n%s", currentOperation, log.getStackTrace()));
+            } else {
+                LOGGER.warning(format("Slow operation detected: %s (%d invocations)%n%s", currentOperation, totalInvocations,
+                        (totalInvocations % FULL_LOG_FREQUENCY == 0) ? log.getStackTrace() : log.getShortStackTrace()));
+            }
+        }
+    }
+
+    private SlowOperationLog getOrCreateSlowOperationLog(OperationRunner operationRunner, final Object operation) {
+        String prefix = "";
+        for (StackTraceElement stackTraceElement : operationRunner.currentThread().getStackTrace()) {
+            stackTraceStringBuilder.append(prefix).append(stackTraceElement.toString());
+            prefix = "\n";
+        }
+
+        final String stackTraceString = stackTraceStringBuilder.toString();
+        stackTraceStringBuilder.setLength(0);
+
+        return ConcurrencyUtil.getOrPutIfAbsent(slowOperationLogs, stackTraceString.hashCode(),
+                new ConstructorFunction<Integer, SlowOperationLog>() {
+                    @Override
+                    public SlowOperationLog createNew(Integer key) {
+                        return new SlowOperationLog(stackTraceString, operation);
+                    }
+                });
+    }
+
+    private boolean purge(long nowNanos, long lastLogPurge) {
+        if (nowNanos - lastLogPurge > logPurgeIntervalNanos) {
+            for (SlowOperationLog log : slowOperationLogs.values()) {
+                log.purgeInvocations(nowNanos, logRetentionNanos);
+                if (log.isEmpty()) {
+                    slowOperationLogs.remove(log.getStackTrace().hashCode());
+                }
+            }
+            return true;
+        }
+        return false;
+    }
```

Comment:
I personally prefer jumping out of a method as soon as you see there is nothing to do. This reduces the amount of nesting.

It will also reduce the npath complexity of your method.

```
  private boolean purge(long nowNanos, long lastLogPurge) {
        if (nowNanos - lastLogPurge <= logPurgeIntervalNanos) {
            return false;
        }
        
       for (SlowOperationLog log : slowOperationLogs.values()) {
                log.purgeInvocations(nowNanos, logRetentionNanos);
                if (log.isEmpty()) {
                    slowOperationLogs.remove(log.getStackTrace().hashCode());
                }
         }           
 }
```